gdal/ogr/ogrsf_frmts/cad/libopencad/cadgeometry.cpp

```diff
@@ -0,0 +1,1159 @@
+
+    matrix[6] = vector.getX() * a00 + vector.getY() * a10 + a20;
+    matrix[7] = vector.getX() * a01 + vector.getY() * a11 + a21;
+    matrix[8] = vector.getX() * a02 + vector.getY() * a12 + a22;
+}
+
+void Matrix::rotate( double rotation )
+{
+    double              s                                                                                        = sin( rotation ), c = cos(
+            rotation ), a00                                                                                      = matrix[0], a01 = matrix[1], a02 = matrix[2], a10 = matrix[3], a11 = matrix[4], a12 = matrix[5];
+
+    matrix[0] = c * a00 + s * a10;
+    matrix[1] = c * a01 + s * a11;
+    matrix[2] = c * a02 + s * a12;
+
+    matrix[3] = c * a10 - s * a00;
+    matrix[4] = c * a11 - s * a01;
+    matrix[5] = c * a12 - s * a02;
+}
+
+void Matrix::scale( const CADVector& vector )
+{
+    matrix[0] *= vector.getX();
+    matrix[1] *= vector.getX();
+    matrix[2] *= vector.getX();
+    matrix[3] *= vector.getY();
+    matrix[4] *= vector.getY();
+    matrix[5] *= vector.getY();
+}
+
+CADVector Matrix::multiply( const CADVector& vector ) const
+{
+    CADVector out;
+    out.setX( vector.getX() * matrix[0] + vector.getY() * matrix[1] + vector.getZ() * matrix[2] );
+    out.setY( vector.getX() * matrix[3] + vector.getY() * matrix[4] + vector.getZ() * matrix[5] );
+    out.setZ( vector.getX() * matrix[6] + vector.getY() * matrix[7] + vector.getZ() * matrix[8] );
+    return out;
+}
+
+//------------------------------------------------------------------------------
+// CADGeometry
+//------------------------------------------------------------------------------
+
+CADGeometry::CADGeometry() : geometryType( UNDEFINED ), thickness( 0 )
+{
+
+}
+
+CADGeometry::~CADGeometry()
+{
+
+}
+
+CADGeometry::GeometryType CADGeometry::getType() const
+{
+    return geometryType;
+}
+
+double CADGeometry::getThickness() const
+{
+    return thickness;
+}
+
+void CADGeometry::setThickness( double thicknes )
+{
+    thickness = thicknes;
+}
+
+RGBColor CADGeometry::getColor() const
+{
+    return geometry_color;
+}
+
+void CADGeometry::setColor( RGBColor color )
+{
+    geometry_color = color;
+}
+
+vector<string> CADGeometry::getEED() const
+{
+    return asEED;
+}
+
+void CADGeometry::setEED( vector<string> eed )
+{
+    asEED = eed;
+}
+
+vector<CADAttrib> CADGeometry::getBlockAttributes() const
+{
+    return blockAttributes;
+}
+
+void CADGeometry::setBlockAttributes( const vector<CADAttrib>& data )
+{
+    blockAttributes = data;
+}
+
+//------------------------------------------------------------------------------
+// CADUnknown
+//------------------------------------------------------------------------------
+CADUnknown::CADUnknown()
+{
+}
+
+void CADUnknown::transform( const Matrix& /*matrix*/)
+{
+}
+
+void CADUnknown::print() const
+{
+    cout << "|---------Unhandled---------|\n" << endl;
```

Comment:
Why two new lines?  If you really want 2, then drop the endl and do `\n\n";`

If you really need to flush, explain why.